### ISSUE TITLE
H5Store passes H5StoreAlreadyOpen error which may lead to misleading behavior

### ISSUE THREAD

**Issue by klywang** [glotzerlab/signac#284]
### Description

Reference to the following commet [[x]](https://github.com/glotzerlab/signac/pull/278#issuecomment-581163649) in PR #278. 
The following code for the context manager in `H5Store` passes `H5StoreAlreadyOpen` errors:
https://github.com/glotzerlab/signac/blob/569b8e5957ea07f00528ae3cfd1392f1c10d1043/signac/core/h5store.py#L317-L322

This may be problematic since it gives users the appearance of opening two context managers with different modes in parallel. In reality, any arguments in the second context manager are ignored.

### To reproduce

The following test is a good example of misleading behavior:
https://github.com/glotzerlab/signac/blob/569b8e5957ea07f00528ae3cfd1392f1c10d1043/tests/test_h5store.py#L117-L127

If you run the line,  `pytest tests/test_h5store.py -k "TestH5StoreOpen"` from the `master` branch, all tests pass. The test passes if you try to write to the h5py read-only file:
```python
    def test_open_write_and_read_only(self):
        with self.open_h5store(mode='w') as h5s_w:
            with self.open_h5store(mode='r') as h5s_r:
                assert 'foo' not in h5s_r
                assert 'foo' not in h5s_w

                h5s_w['foo'] = 'bar'
                h5s_r['foo2'] = 'bar'
                assert 'foo' in h5s_r
                self.assertEqual(h5s_r['foo'], 'bar')
                assert 'foo' in h5s_w
                self.assertEqual(h5s_r['foo'], 'bar')
```
This test passes, but should not.

Adding a lone context manager with mode 'r' will cause the test to fail:
```python
    def test_open_write_and_read_only(self):
        with self.open_h5store(mode='w') as h5s_w:
            with self.open_h5store(mode='r') as h5s_r:
                assert 'foo' not in h5s_r
                assert 'foo' not in h5s_w

                h5s_w['foo'] = 'bar'
                h5s_r['foo2'] = 'bar'
                assert 'foo' in h5s_r
                self.assertEqual(h5s_r['foo'], 'bar')
                assert 'foo' in h5s_w
                self.assertEqual(h5s_r['foo'], 'bar')

        with self.open_h5store(mode='r') as h5s_r:
            h5s_r['foo3'] = 'bar'
```
### Error output
The test will fail at the line , ` h5s_r['foo3'] = 'bar'` due to the following error:
```bash
E   ValueError: Unable to create dataset (no write intent on file)
```
This is likely due to the fact that `H5StoreAlreadyOpen` error causes the context manager to pass any arguments in the second context manager on line 119. This is misleading and it may be best if this behavior is not allowed at all.

**Potentially unrelated issue:** This was discovered while adding new tests for the project and job interface. In some instances, `H5StoreManager` behaves differently as a context manager than `H5Store`. The differences only come through for more advanced tests/usages (like this issue). The project data interface fails the unmodified version of this test due to the `H5StoreAlreadyOpen` error. This should be safer behavior than the behavior observed in `H5Store`. 
The differences in these tests (and some other tests, referenced in this comment [[x]](https://github.com/glotzerlab/signac/pull/278#issuecomment-581163649)) indicates that the context manager for direct instances of `H5Store` and the project data interface may behave very differently, even after the issue addressed in PR #278 was resolved. I'm not sure how similar these two classes were meant to be but I felt that it was worth mentioning.

### System configuration

Please complete the following information:

 - Operating System [e.g. macOS]: Linux Mint (Debian)
 - Version of Python [e.g. 3.7]: 3.8.1
 - Version of signac [e.g. 1.0]: 1.3.0

Or copy & paste the output of: `python -c 'import platform; print(platform.platform()); import sys; print(sys.version); import signac; print(signac.__version__)'`

**Comment by csadorf** [comment]
@klywang Thank you very much for the detailed analysis.

The `H5Store` instance itself has an implicit mode through the `kwargs` argument to the constructor (see [here](https://github.com/glotzerlab/signac/blob/569b8e5957ea07f00528ae3cfd1392f1c10d1043/signac/core/h5store.py#L356)) and will open with that mode when entered as CM. So to my understanding the user cannot open the same store with different mode arguments, since they cannot pass any arguments.

So I'm wondering, is this an issue that can occur in practice or is this more a testing issue? If it's the latter than we should improve our tests, otherwise we should identify and fix the bug.

**Comment by klywang** [comment]
Here is a minimal example of the tests on the `master` branch. This first example passes but should not:
```bash
$ cd /path/to/project/directory/
$ signac shell
>>> file = 'signac_data.h5'
>>> with signac.H5Store(file, mode='w') as h5p_w:
...         with signac.H5Store(file, mode='r') as h5p_r:
...             h5p_w['writer'] = 'hello'
...             h5p_r['reader'] = 'hello'
... 
>>> with signac.H5Store(file, mode='r') as h5p_r:
...         h5p_r['reader']
...         h5p_r['writer']
... 
'hello'
'hello'
```

This second example fails with `ValueError: Unable to create dataset (no write intent on file)`:
```bash
>>> with signac.H5Store(file, mode='r') as h5p_r2:
...         h5p_r2['reader2'] = 'hello'
... 
```

**Comment by bdice** [comment]
@klywang I'm not sure what to expect from the first example, I don't know how multiple file handles on the same file work in practice. This page indicates that the behavior is undefined, in which case we can't do much about it except potentially try to prevent users from opening the same file again: http://docs.h5py.org/en/stable/whatsnew/2.0.html#file-objects-must-be-manually-closed.

The second example (failing to write with a read-only store) seems like expected behavior to me. What did you expect to happen?

**Comment by klywang** [comment]
@bdice I expect the first example to return an `H5StoreAlreadyOpen` error. I don't think users should be allowed to write code like in the first example. There are several tests in `test_hstore.py` which test variations of the first example and expect them to pass. 

The second example is meant to show what should happen when you try to write something in read mode to contrast with the first example, where the user is mislead into believing that they opened the file twice in two different modes.

**Comment by vyasr** [comment]
The core issue confusing @klywang is that `[job|project].data` is a single `H5Store` instance (`project.data` was not, but is being fixed in #278), but Kelly was expecting it to behave more like the class constructor. In practice, the main difference between these two choices is that nothing prevents the user from creating multiple `H5Store` objects that all reference the same file. For instance, this works:

```
>>> with signac.H5Store('signac_data.h5').open(mode='r'):
...     with signac.H5Store('signac_data.h5').open(mode='r'):
...             pass
...
>>>
```

but this does not
```
>>> with j.data.open(mode='r'):
...     with j.data.open(mode='r'):
...             pass
...
Traceback (most recent call last):
  File "<stdin>", line 2, in <module>
  File "/usr/lib/python3.8/site-packages/signac/core/h5store.py", line 357, in open
    return self._open(mode=mode)
  File "/usr/lib/python3.8/site-packages/signac/core/h5store.py", line 330, in _open
    raise H5StoreAlreadyOpenError(self)
signac.core.h5store.H5StoreAlreadyOpenError: H5Store(filename='signac_data.h5')
```

Currently `project.data` does not throw this error, but it will start to after #278 is merged. I think this issue is primarily a result of confusion over how the properties that are `H5Store` objects behave, and is not something we need to change other than perhaps being clearer in the docs that `job.data` is an instance of `H5Store`.

However, I think that @csadorf makes a good point about whether this is a testing bug. Naively, I would not expect the method `TestH5StoreBase.get_h5store` to create a new `H5Store` every time, but rather to only create one for the current test and reuse it for further calls. It's also not intuitive to me that `get_other_h5store` necessarily points to a separate file, rather than being a separate `H5Store` instance pointing to the same underlying file. From looking at the tests, though, I think they are testing the intended behavior, i.e. multiple handles to the same file, so I think it's maybe a question of clarifying the behavior in the naming of the functions. @csadorf @bdice does that sound right to you, or do you think there's more to be done with respect to the tests?

Regarding the question of having multiple open handles to the same HDF5 file, h5py does not prevent you from doing this so I don't think we should either. We could implement a safety check that multiple stores accessing the same file are not open at the same time, but it would require some sort of global (or at least class-level) tracking of all instances and I don't think it's worthwhile. 

Regarding concurrent reads, HDF5 does not really support this. In the very limited case of something exclusively using the C API (not even the C++ API, let alone any higher level calls), it is possible to build a thread safe version that really just serializes reads from multiple threads at the OS level (see [here](https://portal.hdfgroup.org/display/knowledge/Questions+about+thread-safety+and+concurrent+access)). You can also build an MPI safe version that in my understanding basically creates a single global object under the hood and synchronizes some intermediate representation across MPI ranks before writing; h5py has an interface to this behavior. However, nothing in h5py tries to protect the user against concurrent reading/writing, so I don't think we should try to either.

**Comment by klywang** [comment]
@vyasr I think the documentation is clear that `job` is an instance of `H5Store`. I simply did not understand that there is a difference between opening the same file with different `H5Store` objects and opening the same file multile times with the same `H5Store` example. I'm not sure any of this is a documentation issue. Rereading through all of this, I don't think anything needs to be changed and this issue can be closed.
